internal/rbd/replicationcontrollerserver.go

```diff
@@ -0,0 +1,357 @@
+	mirroringMode, err := getMirroringMode(ctx, req.GetParameters())
+	if err != nil {
+		return nil, err
+	}
+
+	mirroringInfo, err := rbdVol.getImageMirroringInfo()
+	if err != nil {
+		util.ErrorLog(ctx, err.Error())
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	if mirroringInfo.State != librbd.MirrorImageEnabled {
+		err = rbdVol.enableImageMirroring(mirroringMode)
+		if err != nil {
+			util.ErrorLog(ctx, err.Error())
+			return nil, status.Error(codes.Internal, err.Error())
+		}
+	}
+	return &replication.EnableVolumeReplicationResponse{}, nil
+}
+
+// DisableVolumeReplication extracts the RBD volume information from the
+// volumeID, If the image is present and the mirroring is enabled on the RBD
+// image it will disable the mirroring.
+func (cs *ControllerServer) DisableVolumeReplication(ctx context.Context,
+	req *replication.DisableVolumeReplicationRequest,
+) (*replication.DisableVolumeReplicationResponse, error) {
+	rbdVol, cr, err := cs.getVolumeFromID(ctx, req.GetVolumeId(), req.GetSecrets())
+	defer cs.cleanup(rbdVol, cr)
+	if err != nil {
+		return nil, err
+	}
+
+	// extract the force option
+	force, err := getForceOption(ctx, req.GetParameters())
+	if err != nil {
+		return nil, err
+	}
+
+	mirroringInfo, err := rbdVol.getImageMirroringInfo()
+	if err != nil {
+		util.ErrorLog(ctx, err.Error())
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	switch mirroringInfo.State {
+	case librbd.MirrorImageEnabled:
+		if !force && !mirroringInfo.Primary {
+			return nil, status.Error(codes.InvalidArgument, "image is in non-primary state")
+		}
+		err = rbdVol.disableImageMirroring(force)
+		if err != nil {
+			util.ErrorLog(ctx, err.Error())
+			return nil, status.Error(codes.Internal, err.Error())
+		}
+		// TODO: check do we need to return abort after disabling the mirroring
+		return &replication.DisableVolumeReplicationResponse{}, nil
+	case librbd.MirrorImageDisabled:
+		// image is already in disabled state
+	default:
+		// TODO: use string instead of int for returning valid error message
+		return nil, status.Errorf(codes.InvalidArgument, "image is in %d Mode", mirroringInfo.State)
+	}
+
+	return &replication.DisableVolumeReplicationResponse{}, nil
+}
+
+// PromoteVolume extracts the RBD volume information from the volumeID, If the
+// image is present, mirroring is enabled and the image is in demoted state it
+// will promote the volume as primary.
+// If the image is already primary it will return success.
+func (cs *ControllerServer) PromoteVolume(ctx context.Context,
+	req *replication.PromoteVolumeRequest,
+) (*replication.PromoteVolumeResponse, error) {
+	rbdVol, cr, err := cs.getVolumeFromID(ctx, req.GetVolumeId(), req.GetSecrets())
+	defer cs.cleanup(rbdVol, cr)
+	if err != nil {
+		return nil, err
+	}
+
+	// extract the force option
+	force, err := getForceOption(ctx, req.GetParameters())
+	if err != nil {
+		return nil, err
+	}
+
+	mirroringInfo, err := rbdVol.getImageMirroringInfo()
```

Comment:
Should you worry about multiple requests running concurrently? I don't see the usual per-image handling lock. At the very least, these should be idempotent so that if you attempt to promote a volume that has already been promoted, you don't get an error. 

Technically I guess this comment applies throughout.

Reply:
Good point @dillaman .  The mirroring operations can run in parallel. The only lock taken here is to retrieve volume from the volumeID which is not enough I feel.

Reply:
currently taking a lock on the volumeID (encoded volumeID) in all the requests. 

@humblec lock is not only to retrieve the volume it will be released at the last after completing all the mirroring operations. if the lock exists on a volumeID another request will try to get the lock it will fail, return operation already exists error message.

Reply:
@humblec do you have any suggestion on the locking what extra lock we need to take?

Reply:
@Madhu-1 Ah, so the lock is now stashed in the `rbdVol` object and the `cleanup` method will release it?

Reply:
the lock is in [ControllerServer structure](https://github.com/ceph/ceph-csi/blob/devel/internal/rbd/controllerserver.go#L53)  as its a global one and the cleanup method will release based on the volumeID

Reply:
@Madhu-1 the snapshot restore operation lock is free from parent volume operation or iow, the snapshot restore operation is performed without a lock on parent volume atm, Isnt it ? if thats the case, I am wondering while  a mirroring operation on parent volume and snapshot restore operation happens ( in one of the snapshot of this volume) in parallel, is there a possibility of race or weird outcome?

Reply:
> @Madhu-1 the snapshot restore operation lock is free from parent volume operation or iow, the snapshot restore operation is performed without a lock on parent volume atm, Isnt it ?

we are taking a lock on the parent volume ID during SnapshotCreate operation here https://github.com/ceph/ceph-csi/blob/devel/internal/rbd/controllerserver.go#L783-L788. During Restore operation we are restoring from a clone that is created on the parent volume During the SnapshotCreate operation.

Yes, it's already a clone do we need to worry about the lock on parent volume Here?

 if thats the case, I am wondering while a mirroring operation on parent volume and snapshot restore operation happens ( in one of the snapshot of this volume) in parallel, is there a possibility of race or weird outcome?

I don't see an issue here. Please let me know if have missed a corner case here.

Reply:
May be @dillaman would be the right person to validate above scenario.